### PR TITLE
Standardize TypeScript and lint configuration

### PR DESCRIPTION
**Pull request by rekmarks**
- TypeScript configuration
  - Update target from `es5` to `ES2017`
  - Add source maps and other config options per other repositories
  - Add multiple `tsconfig` files to support separate bundling of source and test files
- Move to `src`/`dist` file structure
- Update `eslint` config and dependencies
  - Manually re-type a couple of things to fix lint errors
- Run `prettier` on everything

No functional changes.

### CODE REVIEW

**Review by Gudahtt — CHANGES_REQUESTED** [review]
I've removed `eccrypto-lite` in a separate PR: https://github.com/MetaMask/eth-sig-util/pull/116

Once that is merged, could you rebase this to fix the conflicts?

**Comment by rekmarks** [comment]
@Gudahtt done!

File: package.json [hunk 1]
```diff
@@ -42,15 +41,17 @@
...
     "@types/node": "^10.17.13",
-    "@typescript-eslint/eslint-plugin": "^2.20.0",
-    "@typescript-eslint/parser": "^2.20.0",
+    "@typescript-eslint/eslint-plugin": "^4.12.0",
+    "@typescript-eslint/parser": "^4.12.0",
     "browserify": "^16.5.1",
     "eslint": "^6.8.0",
-    "eslint-plugin-import": "^2.20.1",
+    "eslint-plugin-import": "^2.22.1",
+    "eslint-plugin-json": "^2.1.2",
```
Comment (Gudahtt) [inline package.json:51]: We should consider dropping this in favour of `prettier`. That's what we did on the extension.

File: .gitignore [hunk 2]
```diff
@@ -1,12 +1,9 @@
-test/bundle.js
```
Comment (Gudahtt) [inline .gitignore:1]: This still needs to be ignored until #115 is merged. We could merge that first though.

File: package.json (hunk 1 above)
Comment (rekmarks) [reply package.json:51]: I'd like to see us add `prettier` to our base `eslint` config first, and add the `prettier` plugin as a peerDependency. In fact, I was discussing this with @brad-decker earlier, and I also suggested that we do this: https://github.com/MetaMask/eslint-config/issues/104

Comment (Gudahtt) [reply package.json:46]: Oh I meant specifically for linting `.json` files. It was added to the extension to lint JSON before it was used for JavaScript.

But sure, we can wait. That works too.

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Pull request merged**